File: crates/dewasm-backend-perl/units/wasi/fd_seek.pl
# requires: memory/i64_store, rt/s64
use Errno ();

sub wasi_fd_seek {
    my ($self, $fd, $offset, $whence, $out_ptr) = @_;
    my $e = $self->{fds}{$fd};
    return ERRNO_BADF if !defined($e) || $e->{dir};
    return ERRNO_SPIPE if defined $e->{std};
    return ERRNO_NOTCAPABLE unless $self->{meta}{$fd}[0] & RIGHTS_FD_SEEK;
    return ERRNO_INVAL if $whence > 2;  # WASI whence 0/1/2 = sysseek SET/CUR/END
    my $pos = sysseek($e->{fh}, Rt::s64($offset), $whence);
    if (!defined $pos) {
        # A negative resulting offset is EINVAL, which WASI reports as
        # INVAL (not the generic IO used for other seek failures).
        return 0 + $! == Errno::EINVAL() ? ERRNO_INVAL : ERRNO_IO;
    }
    $self->{memory}->i64_store($out_ptr, $pos);
    return ERRNO_SUCCESS;
}

File: crates/dewasm-backend-ruby/units/wasi/fd_seek.rb
# requires: memory/ids, rt/s64, rt/m64, wasi/rights
def wasi_fd_seek(fd, offset, whence, out_ptr)
  io = @fds[fd]
  return ERRNO_BADF if io.nil? || io.is_a?(WasiDir)
  return ERRNO_SPIPE if @std_ios.include?(io)
  return ERRNO_NOTCAPABLE unless fd_has_right?(fd, RIGHT_FD_SEEK)
  mode = [IO::SEEK_SET, IO::SEEK_CUR, IO::SEEK_END][whence]
  return ERRNO_INVAL unless mode
  io.seek(Rt.s64(offset), mode)
  @memory.ids(out_ptr, Rt.m64(io.tell))
  ERRNO_SUCCESS
rescue Errno::EINVAL
  # Seeking before byte 0 raises EINVAL; surface it precisely rather than folding it into the generic ERRNO_IO.
  ERRNO_INVAL
rescue SystemCallError
  ERRNO_IO
end
